### PR TITLE
decomp: `hint-control` | `menu` | `default-menu` (as much as possible)

### PR DESCRIPTION
**Pull request by xTVaser**
Up for visibility, starting to change / find potential issues with a variety of things once touching `default-menu`, so it might be useful for others to see.

### CODE REVIEW

File: decompiler/config/all-types.gc [hunk 1]
```diff
@@ -16122,12 +16125,12 @@
...
 ;; - Functions
 
-(define-extern auto-save-command function)
+(define-extern auto-save-command (function symbol int int process-tree none)) ;; TODO - not confirmed
 (define-extern auto-save-init-by-other function)
 (define-extern progress-allowed? (function symbol))
 (define-extern print-game-text (function string font-context symbol int int float)) ; TODO decomp error, this seems correct though
 (define-extern get-aspect-ratio (function symbol))
-(define-extern get-task-status function)
+(define-extern get-task-status (function int uint64)) ;; TODO - not confirmed
```
Comment (water111) [inline decompiler/config/all-types.gc:16133]: should return just a `uint`

File: goal_src/engine/debug/menu.gc [hunk 2]
```diff
@@ -5,3 +5,1968 @@
...
+     ((and
+       (< (the-as int (-> arg0 fval)) 0)
+       (< -100 (the-as int (-> arg0 fval)))
+       )
+      (let ((s5-2 format)
+            (a0-8 (clear (-> arg0 display-str)))
+            (a1-2 "-0.~1d")
+            (v1-8 (abs (the-as int (-> arg0 fval))))
+            )
+       (s5-2 a0-8 a1-2 (/ (mod v1-8 100) 10))
```
Comment (water111) [inline goal_src/engine/debug/menu.gc:293]: hmm, this looks weird.  I'd like to take a look at how the decompiler handles `mod` here first.

File: goal_src/engine/debug/menu.gc [hunk 3]
```diff
@@ -5,3 +5,1968 @@
...
+    0
+    )
+   )
+  (when arg1
+   (let ((t9-0 arg1))
+    (-> arg0 id)
+    3
+    (-> arg0 fval)
+    (-> arg0 fval)
+    (set! (-> arg0 fval) (the-as float (t9-0)))
```
Comment (water111) [inline goal_src/engine/debug/menu.gc:349]: I think this is a funciton call that's missing arguments.

File: goal_src/engine/debug/menu.gc [hunk 4]
```diff
@@ -5,3 +5,1968 @@
...
+   (set! (-> arg0 fval) arg4)
+   (set! (-> arg0 fval) 0.0)
+   )
+  (when arg1
+   (let ((t9-0 arg1))
+    (-> arg0 id)
+    3
+    (-> arg0 fval)
+    (-> arg0 fval)
+    (set! (-> arg0 fval) (the float (t9-0)))
```
Comment (water111) [inline goal_src/engine/debug/menu.gc:385]: I think this is also missing arguments

File: goal_src/engine/debug/menu.gc [hunk 5]
```diff
@@ -5,3 +5,1968 @@
...
+;; definition for function debug-menu-context-release-joypad
+(defun debug-menu-context-release-joypad ((arg0 debug-menu-context))
+  (set! (-> arg0 joypad-func) #f)
+  (set! (-> arg0 joypad-item) #f)
+  #f
+  )
+
+;; definition for function debug-menu-item-get-max-width
+(defun debug-menu-item-get-max-width ((arg0 debug-menu-item) (arg1 debug-menu))
+  (local-vars (v0-1 int))
```
Comment (water111) [inline goal_src/engine/debug/menu.gc:453]: the decompiler should simplify this and I don't know why it doesn't

File: decompiler/config/all-types.gc [hunk 6]
```diff
@@ -18792,178 +18798,166 @@
...
+(define-extern debug-menu-render (function debug-menu int int int int debug-menu))
+(define-extern debug-menu-item-render (function debug-menu-item int int int symbol debug-menu-item))
+(define-extern debug-menu-item-submenu-render (function debug-menu-item-submenu int int int symbol debug-menu-item-submenu))
+(define-extern debug-menu-item-function-render (function debug-menu-item-function int int int symbol debug-menu-item-function))
+(define-extern debug-menu-item-flag-render (function debug-menu-item-flag int int int symbol debug-menu-item-flag))
+(define-extern debug-menu-item-var-render (function debug-menu-item-var int int int symbol debug-menu-item-var))
+(define-extern debug-menu-make-from-template (function debug-menu-context pair debug-menu-node))
+(define-extern debug-menu-append-item (function debug-menu debug-menu-item debug-menu-item))
+(define-extern debug-menu-context-set-root-menu (function debug-menu-context debug-menu debug-menu-context))
+(define-extern debug-menu-func-decode (function object symbol))
```
Comment (water111) [inline decompiler/config/all-types.gc:18952]: should return a function I think?

File: goal_src/engine/debug/menu.gc [hunk 7]
```diff
@@ -5,3 +5,1968 @@
...
+           )
+    (let ((s3-0 (-> (the-as debug-menu s4-0) items))
+          (s4-1 (car arg1))
+          )
+     (set! arg1 (cdr arg1))
+     (let ((s5-0 (car s3-0)))
+      (while (not (null? s3-0))
+       (if
+        (string= (the-as string s4-1) (-> (the-as debug-menu-item s5-0) name))
+        (begin
```
Comment (water111) [inline goal_src/engine/debug/menu.gc:899]: if with a begin in it is something the decompiler should avoid... I need to fix this

File: goal_src/engine/debug/menu.gc [hunk 8]
```diff
@@ -5,3 +5,1968 @@
...
+      (set! (-> v1-16 origin y) (the float a0-18))
+      )
+     (set! sv-16 (-> *display* frames (-> *display* on-screen) frame debug-buf))
+     (set! sv-32 (-> sv-16 base))
+     (let ((t9-2 draw-string)
+           (a0-22 ">")
+           (a1-6 sv-16)
+           )
+      (-> arg0 context font)
+      (t9-2 a0-22 a1-6)
```
Comment (water111) [inline goal_src/engine/debug/menu.gc:1236]: function missing arguments here

File: goal_src/engine/debug/menu.gc [hunk 9]
```diff
@@ -5,3 +5,1968 @@
...
+        )
+       )
+      (set! a0-1 a3-0)
+      (set! a3-0 (cdr a3-0))
+      )
+     )
+    (label cfg-7)
+    (if (null? v1-4)
+     (begin
+      (format 0 "ERROR: Couldn't find selected item in menu.~%")
```
Comment (water111) [inline goal_src/engine/debug/menu.gc:1317]: decompiler should do a nicer job here

File: goal_src/engine/debug/menu.gc [hunk 10]
```diff
@@ -5,3 +5,1968 @@
...
+(defun
+  debug-menu-item-function-msg
+  ((arg0 debug-menu-item-function) (arg1 symbol))
+  (cond
+   ((= arg1 4)
+    (cond
+     ((-> arg0 activate-func)
+      (let ((t9-0 (-> arg0 activate-func)))
+       (-> arg0 id)
+       (if (t9-0)
```
Comment (water111) [inline goal_src/engine/debug/menu.gc:1456]: missing argument here

File: goal_src/engine/debug/menu.gc [hunk 11]
```diff
@@ -5,3 +5,1968 @@
...
+;; definition for function debug-menu-item-flag-msg
+(defun debug-menu-item-flag-msg ((arg0 debug-menu-item-flag) (arg1 int))
+  (cond
+   ((= arg1 4)
+    (when (-> arg0 activate-func)
+     (let ((t9-0 (-> arg0 activate-func))
+           (a0-1 (-> arg0 id))
+           )
+      4
+      (set! (-> arg0 is-on) (the-as symbol (t9-0 a0-1)))
```
Comment (water111) [inline goal_src/engine/debug/menu.gc:1484]: likely missing argument here

File: goal_src/engine/debug/menu.gc [hunk 12]
```diff
@@ -5,3 +5,1968 @@
...
+     (debug-menu-context-send-msg a0-2 3 2)
+     )
+    )
+   ((or (= arg1 3) (= arg1 1))
+    (when (-> arg0 activate-func)
+     (let ((t9-2 (-> arg0 activate-func))
+           (a0-5 (-> arg0 id))
+           )
+      3
+      (set! (-> arg0 is-on) (the-as symbol (t9-2 a0-5)))
```
Comment (water111) [inline goal_src/engine/debug/menu.gc:1497]: missing argument here

File: goal_src/engine/debug/menu.gc [hunk 13]
```diff
@@ -5,3 +5,1968 @@
...
+       (when (-> arg0 factivate-func)
+        (let ((t9-1 (-> arg0 factivate-func))
+              (a0-2 (-> arg0 id))
+              (a1-1 4)
+              (a2-0 (-> arg0 fundo-val))
+              )
+         (-> arg0 fval)
+         (set!
+          (-> arg0 fval)
+          ((the-as (function int float float float) t9-1)
```
Comment (water111) [inline goal_src/engine/debug/menu.gc:1529]: missing argument here

File: goal_src/engine/debug/menu.gc [hunk 14]
```diff
@@ -5,3 +5,1968 @@
...
+  )
+
+;; definition for function debug-menus-active
+(defun debug-menus-active ((arg0 debug-menu-context))
+  (when (not (-> arg0 is-hidden))
+   (cond
+    ((-> arg0 joypad-func)
+     (let ((t9-0 (-> arg0 joypad-func)))
+      (-> arg0 joypad-item)
+      (t9-0)
```
Comment (water111) [inline goal_src/engine/debug/menu.gc:1946]: missing argument

**Review by water111 — COMMENTED** [review]
Nice work, this is a huge file. I had a few comments on function types but it looks very close.

File: goal_src/engine/debug/menu.gc (hunk 8 above)
Comment (xTVaser) [reply goal_src/engine/debug/menu.gc:1236]: hmm this looks like an original bug.  Looks like they meant to use `draw-string-adv` which takes a `font-context` as a third argument.  `draw-string` looks like it only takes 2 arguments (as seen in debug.gc)

**Pull request merged**